Standards/scs-0213-v1-k8s-nodes-anti-affinity.md

```diff
@@ -0,0 +1,123 @@
+decisions about scheduling, event detection and global decisions. Inside the control plane,
+multiple components exist, which can be duplicated and distributed over multiple machines
+inside the cluster. Typically, no user containers are run on these machines in order to
+separate the control plane from the live system.
+
+## Motivation
+
+In a productive environment, the control plane usually runs across multiple machines and
+a cluster usually contains multiple worker nodes in order to provide fault-tolerance and
+high availability.
+
+In order to ensure availability and scaling of workloads, even if some machines in the cluster
+could fail, they should be distributed over multiple nodes on different machines.
+This can be steered with the Affinity or Anti Affinity features, which are separated by
+Kubernetes into two features:
+
+Node Affinity
+The Node Affinity feature allows to match pods according to logical matches of
+key-value-pairs referring to labels of nodes.
+These can be defined with different weights or preferences in order to allow fine-grained
+selection of nodes. The feature works similar to the Kubernetes nodeSelector.
+It is defined in the PodSpec using the nodeAffinity field in the affinity section.
+
+Pod Affinity
+Pod Affinity or Pod Anti Affinity allows the constraint of pod scheduling based on the
+labels of pods already running on a node.
+This means the constraint will match other pods on a node according to their labels key-value-pairs
+and then either schedule the pod to the same (Affinity) or another (Anti Affinity) node.
+This feature is also defined in the PodSpec using the podAffinity field in the affinity section.
+
+Both features allow the usage of "required" or "preferred" keywords, which create
+"hard" or "soft" affinities. By using a hard affinity, a pod would need to be scheduled
+according to the rules set. If this possibility is not given, the pod can't be scheduled.
+A soft affinity would allow scheduling even if the requirements are not fulfilled, but
+they would be preferred if possible.
+
+These features allow efficient resource usage (e.g. by scheduling workloads to evenly
+distribute across nodes) and provide fault-tolerance and therefore high availability.
+But they also require more work during the setup of a service architecture, since nodes
+and pods need to be labelled and described consistently.
+
+In the case of SCS, affinity of the workloads themselves is not relevant, since this
+feature is mostly used by the customers of the providers.
+Instead, the expected standard should enable the Kubernetes cluster to handle Anti Affinity
+rules with a real physical separation as well as distributing the control plane over
+multiple machines in order to provide fault-tolerance during system outages.
+If the control plane survives an outage, a Kubernetes cluster can recover later on.
+
+## Design considerations
+
+SCS plans to require a Hard Anti Affinity and/or Redundancy for the control plane and
```

Comment:
In line 56 you exclude affinity for worker nodes and here you include it. Isn't this contradicting?

Reply:
Maybe its not perfectly formulated. What I meant is that (Anti) Affinity isn't relevant for the SCS internally, since SCS in the end only standardizes the infrastructure, not the workload. But it is nonetheless important to standardize the possibility for (Anti) Affinity usage in the clusters.